rfcs/0000_project_typing_guidelines.md

```diff
@@ -0,0 +1,164 @@
+
+More complex types, like dataclasses and regular classes, are examples of nominal types in Python. Two dataclasses with the same fields but different names would type-check as different types. Classes and dataclasses (which are a convenience for defining certain kinds of classes) are more opaque than anything to do with the fields or methods they have. Different classes are different types and the only way instances of those classes can type check as each other is if one is a subclass of another. 
+
+Python also provides _newtype_s which can be used to define nominal types for basic types that would normally be structurally typed. An example of a newtype would be to define the concept of an integer ID for a user or agent in a system, and ensure that they can't be used in the wrong places.
+
+```python
+UserID = NewType('UserID', int)
+AgentID = NewType('AgentID', int)
+
+# Both UserID and AgentID are integers, but they cannot be swapped.
+def activate_agent(agent: AgentID, actor: UserID) -> None:
+    pass
+
+user_id = UserID(1)
+agent_id = AgentID(2)
+
+activate_agent(agent_id, user_id) # works fine
+# FAILS to typecheck: "UserID" is not "AgentID", etc.
+activate_agent(user_id, agent_id)
+# FAILS to typecheck: "int" is not "UserID"/"AgentID"
+activate_agent(1, 2)
+```
+
+Another example would be to codify the order of a quaternion (since there is disagreement between libraries whether to use WXYZ or XYZW) and then check usages.
+
+Newtypes should be used when **no additional functionality** beyond the underlying type is needed, but metadata about the type needs to be tracked. An example of this would be unsafe and safe strings in a web application. They shouldn’t be confused because they can lead to security vulnerabilities, but with newtypes we can help the author to think about which is being used.
+
+```python
+UnsafeString = NewType('UnsafeString', str)
+SafeString = NewType('SafeString', str)
+
+def render_with_content(template: SafeString, content: SafeString):
+    # This isn't the best way to do this but it's an example
+    return format(template, content)
+
+def sanitize_string(s: UnsafeString) -> SafeString:
+    # do some sanitization
+    return SafeString(new_s)
+
+unsafe_input: UnsafeString # comes from some user input
+template: SafeString = "Hello, {}"
+# FAILS to typecheck: "UnsafeString" is not "SafeString"
+render_with_content(template, unsafe_input)
+
+safe_input = sanitize_string(unsafe_input)
+# FAILS to typecheck: "SafeString" is not "UnsafeString"
+safe_input = sanitize_string(safe_input)
+```
+
+In this example, it becomes more difficult to accidentally pass an unsafe string to a rendering function that could cause security issues, and it’s also difficult to accidentally sanitize a safe string a second time.
+
+### Protocols SHOULD be used to define behaviour that would normally be duck typed
```

Comment:
It could be worth to also link to our other guidance on Protocols here https://thousandbrainsproject.readme.io/docs/code-style-guide#protocols-should-be-preferred-to-document-usage-and-expectations (unless this will be integrated in the same place anyways and the reader has just read that)

Reply:
The plan is to put them in the same place, and since the two documents complement each other, linking between them makes sense.